File: ui/android-app/app/src/test/java/org/aerobag/app/domain/ImageViewportTest.kt
// SPDX-FileCopyrightText: 2026 Aerobag contributors
//
// SPDX-License-Identifier: AGPL-3.0-or-later

package org.aerobag.app.domain

import kotlin.math.abs
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class ImageViewportTest {
    @Test
    fun `initial viewport fits and centers image`() {
        val viewport = createInitialImageViewport(
            imageWidthPx = 1200f,
            imageHeightPx = 800f,
            viewportWidthPx = 900f,
            viewportHeightPx = 700f,
        )
        assertEquals(1f, viewport.zoom)
        assertEquals(0f, viewport.leftPx, 0.001f)
        assertEquals(50f, viewport.topPx, 0.001f)
    }

    @Test
    fun `zoom around point preserves anchor`() {
        val start = createInitialImageViewport(
            imageWidthPx = 1200f,
            imageHeightPx = 800f,
            viewportWidthPx = 900f,
            viewportHeightPx = 700f,
        )
        val anchorX = 300f
        val anchorY = 250f
        val beforeLocalX = anchorX - start.leftPx
        val beforeLocalY = anchorY - start.topPx
        val next = zoomImageAroundPoint(
            state = start,
            anchorX = anchorX,
            anchorY = anchorY,
            nextZoom = 2f,
            imageWidthPx = 1200f,
            imageHeightPx = 800f,
            viewportWidthPx = 900f,
            viewportHeightPx = 700f,
            overscrollPx = 64f,
        )
        val afterLocalX = (anchorX - next.leftPx) / next.zoom
        val afterLocalY = (anchorY - next.topPx) / next.zoom
        assertTrue(abs(afterLocalX - beforeLocalX) < 0.01f)
        assertTrue(abs(afterLocalY - beforeLocalY) < 0.01f)
    }

    @Test
    fun `overscroll is limited to one thumb`() {
        val start = createInitialImageViewport(
            imageWidthPx = 1200f,
            imageHeightPx = 800f,
            viewportWidthPx = 900f,
            viewportHeightPx = 700f,
        )
        val dragged = dragImageViewport(
            state = start,
            dxPx = 600f,
            dyPx = 500f,
            imageWidthPx = 1200f,
            imageHeightPx = 800f,
            viewportWidthPx = 900f,
            viewportHeightPx = 700f,
            overscrollPx = 64f,
        )
        assertTrue(dragged.leftPx <= 64f)
        assertTrue(dragged.topPx <= 64f)
    }
}

File: ui/web-app/src/domain/imageViewport.test.ts
// SPDX-FileCopyrightText: 2026 Aerobag contributors
//
// SPDX-License-Identifier: AGPL-3.0-or-later

import { describe, expect, it } from "vitest";
import conformance from "../generated/uiGeometryConformance.json";
import {
  clampImageViewport,
  createInitialImageViewport,
  dragImageViewport,
  zoomImageAroundPoint,
} from "./imageViewport";

describe("imageViewport", () => {
  it("fits and centers the image initially", () => {
    const viewport = createInitialImageViewport(1200, 800, 900, 700);
    expect(viewport.zoom).toBe(1);
    expect(viewport.left).toBeCloseTo(0);
    expect(viewport.top).toBeCloseTo(50);
  });

  it("preserves the anchor point when zooming", () => {
    const start = createInitialImageViewport(1200, 800, 900, 700);
    const anchor = { x: 300, y: 250 };
    const beforeLocal = {
      x: anchor.x - start.left,
      y: anchor.y - start.top,
    };
    const next = zoomImageAroundPoint(start, anchor.x, anchor.y, 2, 1200, 800, 900, 700, 64);
    const afterLocal = {
      x: (anchor.x - next.left) / next.zoom,
      y: (anchor.y - next.top) / next.zoom,
    };
    expect(afterLocal.x).toBeCloseTo(beforeLocal.x, 4);
    expect(afterLocal.y).toBeCloseTo(beforeLocal.y, 4);
  });

  it("allows one-thumb overscroll but no more", () => {
    const start = createInitialImageViewport(1200, 800, 900, 700);
    const dragged = dragImageViewport(start, 600, 500, 1200, 800, 900, 700, 64);
    const clamped = clampImageViewport(dragged, 1200, 800, 900, 700, 64);
    expect(clamped.left).toBeLessThanOrEqual(64);
    expect(clamped.top).toBeLessThanOrEqual(64);
  });

  it("keeps a real drag range when the fitted image is smaller than the viewport", () => {
    const start = createInitialImageViewport(600, 2400, 1200, 900);
    const leftDragged = dragImageViewport(start, -800, 0, 600, 2400, 1200, 900, 64);
    const rightDragged = dragImageViewport(start, 800, 0, 600, 2400, 1200, 900, 64);
    expect(leftDragged.left).toBeCloseTo(64, 4);
    expect(rightDragged.left).toBeCloseTo(911, 4);
  });

  it("matches core's shared out-of-range zoom clamp vector", () => {
    const vector = conformance.image_clamp;
    expect(clampImageViewport(
      vector.state,
      vector.image_width,
      vector.image_height,
      vector.viewport_width,
      vector.viewport_height,
      vector.overscroll,
    )).toEqual(vector.expected);
  });
});
